### ISSUE TITLE
Agree on formatting style, and then create .clang-format config file to enforce it

### ISSUE THREAD

**Issue by seanm** [libusb/libusb#1442]
We should agree on a formatting style.

Some files intent with tabs, some with spaces.  Some put braces one way, some another.

Once we agree on a style, we can then autoformat using the [clang-format](https://clang.llvm.org/docs/ClangFormat.html) tool. There are even tools that examine a codebase and generate a preliminary clang-format config file based on the predominant style.

**Comment by seanm** [comment]
I started some work here: https://github.com/libusb/libusb/pull/1443

First thing to decide on: tabs vs spaces :)

**Comment by Youw** [comment]
I think we will not be able to get a single `.clang-format` for the entire project. At least not at the beginning.
We need to figure out if it is possible to have several flavours of it at least per-platform.
So we can select "tabs vs spaces" at least per-plaform based on how it is mostly used right now.

**Comment by Youw** [comment]
Second thing: I think we should come up with a `.clang-format`, which, when applied - produces (reasonably) the smallest diff in the existing code.

**Comment by Youw** [comment]
> Why is that desirable?

To preserve git history - it matters:
![image](https://github.com/libusb/libusb/assets/5939659/acec50c8-0e30-4e81-8262-977a99cbe38c)

If we apply `.clang-format` to all existing code - _a lot_ of immediate blame/history will be hidden under something like "applying code style".
If we agree but don't apply format to all existing code - any change to places where something (e.g. tags vs spaces) is different would introduce lots of weird inconsistencies. I've seen that - looks terrible.

> I did that already,

I meant per-backend, or even per-file.

**Comment by seanm** [comment]
> If we apply `.clang-format` to all existing code - _a lot_ of immediate blame/history will be hidden under something like "applying code style".

Which is why `git blame` has `--ignore-rev` and `--ignore-revs-file`. We can create a file listing all revisions that should be ignored in blame and even put that in the `.gitconfig` file.

**Comment by tormodvolden** [comment]
> > Second thing: I think we should come up with a `.clang-format`, which, when applied - produces (reasonably) the smallest diff in the existing code.
> 
> I did that already, using https://github.com/mikr/whatstyle. The result is the first commit here.

When doing this I think some oddballs in os/* should be taken out. Maybe that whole folder.

**Comment by tormodvolden** [comment]
> > We need to figure out if it is possible to have several flavours of it at least per-platform.
> 
> Why is that desirable? Such a small project could surely have a uniform style.

There is code mainly worked on by people deeply rooted in their own platform style. I wouldn't mind having the Windows code Windows style and the openbsd code openbsd style. OTOH we should expect their contributors to be able to adapt to our common style also.

**Comment by seanm** [comment]
> There is code mainly worked on by people deeply rooted in their own platform style. I wouldn't mind having the Windows code Windows style and the openbsd code openbsd style.

So let's see, that makes:

- linux style
- darwin style
- OpenBSD style
- Haiku style
- android style
- SunOS style
- NetBSD style
- core / crossplatform style

All that for a project with a mere 45 source files?

It is possible to have different `.clang-format` files per directory, but a lot of these variants are all together in the `os` folder.

It seems the choices are:

1. forget about clang-format
2. agree on a single libusb style
3. reorganize files into different folders so they can have different styles

I vote for 2 personally. I don't care what the style is. Being able to auto-apply style makes contributing to the project easier, and it makes code reviews easier.

**Comment by Youw** [comment]
Optione 3 is something that hasn't been on the table and looks interesting.
Does git preserves a file history when it is moved to a different folder?

**Comment by seanm** [comment]
>Optione 3 is something that hasn't been on the table and looks interesting.

Perhaps we should consider how many differences there actually are?  If few, then moving files around may be more disruptive than (for example) just standardizing on tabs vs spaces.

> Does git preserves a file history when it is moved to a different folder?

Subversion does, so surely git does too.

**Comment by Youw** [comment]
> Subversion does, so surely git

I wouldn't be so sure - need to check.

**Comment by tormodvolden** [comment]
I would be opposed to option 3, if clang-format cannot do better then I would rather put "// clang-format off" into the odd-ball files, _if_ we want to keep them odd-ball.

> produces (reasonably) the smallest diff in the existing code.

And again I think this exercise needs to be redone ignoring the current oddballs, or all os/* files. I don't think e.g. the haiku files should have a "vote" in our .clang-format specification.

**Comment by tormodvolden** [comment]
To not hurt any feelings you could exclude all os/* :) But I think the `events_*.c threads_*.c linux_*.c` files are pretty consistent with our core files.

**Comment by seanm** [comment]
@tormodvolden ok see new commits, and commit messages.

**Comment by seanm** [comment]
@tormodvolden @Youw I looked at this a bit more just now the result `whatstyle` gives is pretty good. One thing the core files are pretty inconsistent about, that we should decide on, is this setting: https://clang.llvm.org/docs/ClangFormatStyleOptions.html#alignafteropenbracket

preferences? I vote `AlignAfterOpenBracket: true`

**Comment by seanm** [comment]
@tormodvolden @Youw with 1.0.29 just released, I'd really like to get back to this.

What do you guys prefer for https://clang.llvm.org/docs/ClangFormatStyleOptions.html#alignafteropenbracket ?

**Comment by Youw** [comment]
Unfortunately I'm not statistically familiar with libusb code enough. My personal preference is either `BAS_BlockIndent` or `BAS_AlwaysBreak` (in that order), but I think we should choose what is more common for the current code-base, to have fewer changes/diff once we agree on the code-style (and enforce it).

**Comment by seanm** [comment]
> ... but I think we should choose what is more common for the current code-base...

Yes, that's why I ran the `whatstyle` tool to examine the code base and choose matching settings. But sometimes the codebase is 50/50 or 60/40 split, and then we may as well choose ourselves.

**Comment by Mr-Bossman** [comment]
@seanm from what I can tell looking at the core files the most common style is `BAS_DontAlign`, but my preference is `BAS_Align` and more generally the Linux kernel's style guide. Is there a way to just have clang-format for the core files? Or maybe also include `events_*.c`, `threads_*.c`, and `linux_*.c`?

**Comment by seanm** [comment]
OK I played with this more, and here's the result of my current draft .clang-format rules applied to core.c: https://github.com/libusb/libusb/pull/1443/commits/e2a17ce456f08d7685b16d073470ca319191f149

Any objectionable changes in there?  If so, I can try to find a setting to change to our liking...

**Comment by seanm** [comment]
So it's been almost 2 years since I opened this Issue. :(

The lack of automatic formatting results in minor things like this: https://github.com/libusb/libusb/pull/1723/files/b5993c94bf853ffb7c7f1cbcb5a73d7468881047

It's a waste of reviewers' time to notice such things, and it's discouraging to contributors, new ones especially, to have such things pointed out.  It could all be automated.

How can we move forward with this?

**Comment by sonatique** [comment]
Hi @seanm , I am not familiar with how "autoformat using the [clang-format](https://clang.llvm.org/docs/ClangFormat.html) tool." will work in practice for libusb. Say I am on my computer coding my next PR for libusb (say on Windows). Do I need to install tools for the autoformat to work? Or will format be enforced on github?

Could you write 2-3 lines about what would change in the workflow for Windows, Linux, OSX and the and github? Maybe this could help make things clearer for everybody and help promptly adopting your suggestion.

I am really in favor of having a tool for formatting libusb code and I can adapt to any convention.

**Comment by seanm** [comment]
Lots of different code editors support clang-format, but of course not all of them. Some editors will auto-format as you type, others you have to invoke a command like 'auto-format this file'.  The suggested workflow would be to do this before you commit. But there needn't be any obligation for contributors to use clang-format. Everyone could contribute as they do today. Others could invoke clang-format every once in a while to clean up formatting.

What editor do you use?

I'm not familiar with github and its abilities, but I've seen other projects integrate it into git hooks and CI. I don't know how to do that. But perhaps it could be done one day. It would of course first require we have a .clang-format config file.

**Comment by seanm** [comment]
Once we get https://github.com/libusb/libusb/pull/1731 in, we'll have all ***files*** self-consistent in indentation. I think the next thing to tackle is making ***folders*** self-consistent (in indentation).

**Comment by Youw** [comment]
Once we fixup consistency, we would need to have automated check for consistency for each PR.
